pkg/kubelet/kubelet.go

```diff
@@ -2444,31 +2455,75 @@ func (kl *Kubelet) cleanUpContainersInPod(podID types.UID, exitedContainerID str
 	}
 }
 
-// fastStatusUpdateOnce starts a loop that checks the internal node indexer cache for when a CIDR
-// is applied  and tries to update pod CIDR immediately. After pod CIDR is updated it fires off
-// a runtime update and a node status update. Function returns after one successful node status update.
+// fastStatusUpdateOnce starts a loop that checks if the current state of kubelet + container runtime
+// would be able to turn the node ready, and sync the ready state to the API server as soon as possible.
+// Function returns after the node status update after such event, or when the node is already ready.
 // Function is executed only during Kubelet start which improves latency to ready node by updating
-// pod CIDR, runtime status and node statuses ASAP.
+// kubelet state, runtime status and node statuses ASAP.
 func (kl *Kubelet) fastStatusUpdateOnce() {
+	defer func() {
+		kl.containerRuntimeReadyExpected = true
+	}()
+
 	ctx := context.Background()
-	for {
-		time.Sleep(100 * time.Millisecond)
-		node, err := kl.GetNode()
+
+	nodeIsReady := func(node *v1.Node) bool {
+		for _, c := range node.Status.Conditions {
+			if c.Type == v1.NodeReady {
+				return c.Status == v1.ConditionTrue
+			}
+		}
+
+		klog.ErrorS(nil, "Node does not have NodeReady condition", "node", node)
+		return false
+	}
+
+	var originalNode, node *v1.Node
+
+	start := kl.clock.Now()
+
+	// Sleep() is here so the delay only applies on retry.
+	for ; ; kl.clock.Sleep(100 * time.Millisecond) {
+
+		if kl.clock.Since(start) > nodeReadyGracePeriod {
+			klog.ErrorS(nil, "Node not becoming ready in time after startup")
+			return
+		}
+
+		var err error
+		originalNode, err = kl.GetNode()
 		if err != nil {
 			klog.ErrorS(err, "Error getting node")
 			continue
 		}
-		if len(node.Spec.PodCIDRs) != 0 {
-			podCIDRs := strings.Join(node.Spec.PodCIDRs, ",")
-			if _, err := kl.updatePodCIDR(ctx, podCIDRs); err != nil {
-				klog.ErrorS(err, "Pod CIDR update failed", "CIDR", podCIDRs)
-				continue
-			}
```

Comment:
for other reviewers, I don't think we should do this and that it really has to be removed.

It seems to be an optimization for kubenet (that no longer exist) and seems that is not in use anymore, checking at containerd code the PodCIDR passed through the runtime is only used for the cni_template feature https://github.com/containerd/containerd/commit/b2099c206107a16d1a48942de699a1f1cac356de that AFAIK is only used for testing and CRIO-O doesn't seem to use it at all https://github.com/cri-o/cri-o/search?q=podcidr

Also, and more important, the CNI should reconcile against the node Object and not do a one shot configuration, the kubelet may restart and change the assigned PodCIDR

Reply:
Checking the original PR https://github.com/kubernetes/kubernetes/pull/67031 it seems we can also get rid of some mutex

Reply:
syncNodeStatus is still called from both fastStatusUpdateOnce and the regular jitter.

updatePodCIDR is still called from both jitter->syncNodeStatus->updateNode and fastStatusUpdateOnce->updateNode, unless we need to add a mutex on updateNode.

updateRuntimeUp is still called from both fastStatusUpdateOnce and the regular jitter.

So probably nothing can be removed, if not adding more.

As for adding more, I'm not sure if Kubelet.patchNodeStatus needs a new mutex, and if so, does it need to cover nodeutil.patchNodeStatus or not.

Reply:
🤕 😵‍💫

Reply:
Is `kl.setNodeStatus(node)` thread safe or does it need a new mutex?

Reply:
I'm actually thinking if I should get rid of the several locks and acquire syncNodeStatusMux throughout the whole loop body.

Reply:
Example of replacing with less granular mutex: https://github.com/kubernetes/kubernetes/pull/112618/commits/203989aca82096d5734c627d0fdb3669bc07a587

Reply:
I feel https://github.com/kubernetes/kubernetes/commit/203989aca82096d5734c627d0fdb3669bc07a587 is much more difficult to read and easier to go wrong

Reply:
Okay let's keep the granular mutexes then. Can someone help determining if `kl.setNodeStatus(node)` needs mutex? It's too many functions called there.

It appears at least some of them include their own lock, for example `kl.recordNodeSchedulableEvent` in https://github.com/kubernetes/kubernetes/blob/64af1adaceba4db8d0efdb91453bce7073973771/pkg/kubelet/kubelet_node_status.go#L574-L576 - so it's likely good.

However there's also things like `kl.cloudResourceSyncManager.NodeAddresses` so I wonder if there's any convention such as "NodeStatusFuncs are expected to be thread-safe, and if they're not it's their own fault".

Reply:
@aojea what do you think about https://github.com/kubernetes/kubernetes/pull/112618/commits/4ac63895ed0b099d576b63ce4d9be8b60115e1ee then?

Reply:
with this we'd better move fastStatusUpdate into kubelet_node_status.go and it's really a "lightweight syncNodeStatus" that can be called frequently without putting a load on the apiserver.

Reply:
if you want this into this release I will postpone the mutex optimization, I don't think these days any of the current people in the project has enough context, at least I can't see the previous authors very active last years ... and without context I think we should be cautious

Reply:
Let's try https://github.com/kubernetes/kubernetes/pull/112618/commits/b4ab91601728a452773b3facfec77386e672dcb1

This restructured the original commit to allow easier removal later on. I think it's actually more clear than the original one.

Removal will be like https://github.com/kubernetes/kubernetes/compare/b4ab91601728a452773b3facfec77386e672dcb1..4ac63895ed0b099d576b63ce4d9be8b60115e1ee then needStatusReport/markStatusReport can be merged into the caller if no mux is necessary.

Reply:
setNodeStatus is no longer a concern since I've wrapped the whole `fastNodeStatusUpdate` in the `syncNodeStatusMux`.

Removing excessive locks would be the next step.

Reply:
+1 on removing the excessive locks in a followup PR

Reply:
+1 on removing the locks in a followup PR to keep this change more self-contained.